### DISCUSSION TITLE
FR: allow package in nested folder, defined in `.spi.yml`

### DISCUSSION THREAD

**Discussion by mesqueeb** [SwiftPackageIndex/SwiftPackageIndex-Server#3197] [Ideas]
### The problem

When working on any package, my usual set up is mono-repository with:
1. the swift package (eg. `/MyPackage`)
2. an example Xcode project for development purposes (eg. `/MySampleApp`)

When these sit next to each other in each their own folder, it will not satisfy SPI's requirement:
> The packages must all contain a valid Package.swift file in the root folder.

When I move the `Package.swift` to sit at the repository root however, I have to jump through annoying hoops to make that package importable into the example project that also sits in that same mono-repo, because of this warning in Xcode:
> The selected package cannot be a direct ancestor of the project.

<img src="https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/assets/3253920/a691ae5b-b959-4e3a-8d18-9a21e0c483b6" width="200" />

### Proposed solution

The solution is simply to have 2 sibling folders in the monorepo `/MySampleApp` and `/MyPackage` and then if we can just let SPI know via `.spi.yml` it would solve a lot of my problems.

Eg.
```yml
version: 1
builder:
  configs:
    - package_path: ./MyPackage
    - documentation_targets: [MyPackage]
```

**Comment by daveverwer** [comment]
I may be mistaken here, but I thought SwiftPM itself needed the manifest file to be located in the root directory. I presume it does, but does a package configured as you say work with SwiftPM? Do you have an example package configured this way so I can test it without needing to make one?

**Reply by mesqueeb** [reply]
@daveverwer I'm not 100% sure what you mean, exactly. My proposed solution is for a new variable `package_path` that we can define in `.spi.yml` to allow the `Package.swift` file to sit in a nested folder. I'm OK with `.spi.yml` staying in the root of course.

Of course my proposed solution doesn't work with SwiftPM, it's a feature request. 😅  

An example of my use case however, you can find here: https://github.com/mesqueeb/MicroMaxOnAppleSilicon

if you just open the root `MicroMaxOnAppleSilicon.xcworkspace` in Xcode you will see I have one Xcode project but my Swift Package itself has its `Package.swift` on the repo root, so it's extremely tricky how I had to set up the sample project to be able to "import" the swift package that is sitting at the monorepo root...

Here I try to explain the setup via this screenshot:

<img width="1230" alt="my hoops" src="https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/assets/3253920/7f466b0a-2ec9-46df-830e-3ffe28a88fa4">

All of this will be immensely simplified IF spi allows the `Package.swift` file to sit in a nested folder instead of the repository root folder! 😂

**Comment by daveverwer** [comment]
This is what I was asking for. Thank you. What I was asking was whether the approach you're using here is compatible with SwiftPM, meaning can I use this package from another project? If not, even if it works inside your app, it won't be useful in the index.

I'll check and get back to you.

**Comment by daveverwer** [comment]
I just took a quick look at your package and I see that it currently has the Package.swift manifest in the root, presumably moved there to satisfy our requirement for it to be in the root. My apologies, I thought you were referring to a package you could not get into the index because of this requirement.

So, instead. I moved the Package.swift manifest in my test package and was unable to import it in either an Xcode project or a Package project. The error is fairly clear:

![Screenshot 2024-07-09 at 14 41 43@2x](https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/assets/5180/1ac15c9a-2eb9-41cb-832a-0aebd8115045)

![Screenshot 2024-07-09 at 14 45 42@2x](https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/assets/5180/6fe9b405-176c-4ff5-a6f9-35f4f64ae63b)

I didn’t think that rule was a SPI-specific rule and as far as I can tell I was correct. SwiftPM does not support packages that do not have their manifest in the root of the project.

**Comment by daveverwer** [comment]
Tracking a potential improvement to SwiftPM documentation to clarify this here: https://github.com/swiftlang/swift-org-website/pull/756

**Comment by mesqueeb** [comment]
@daveverwer I deleted my last reply because I found a **much better work around** for my issue.

1. Set up 2 folders, one for the Swift Package and one for the Xcode project
2. Save a new copy of the Swift Package `Package.swift` file on the repository root, and just tweak the target `path` properties to point to the nested folder.

- Now the Xcode project can easily import the Swift Package because it uses the one inside the nested folder
- And exposing the Swift Package to SPI/SPM will still work correctly because we also have a `Package.swift` located at the repository root
